modules/dlm/src/main/java/org/elasticsearch/dlm/DataLifecycleService.java

```diff
@@ -331,6 +431,166 @@ public void onFailure(Exception e) {
+     * cluster state indicating that force merge has begun. And once the request has completed successfully it writes a piece of custom
+     * metadata in the cluster state indicating that the force merge has completed. The listener is notified when the forcemerge request
+     * has completed successfully, or when the forcemerge fails or the write of the begin token to the cluster state fails. The success or
+     * failure of the write of the completed token to the cluster state is only logged.
+     */
+    private void forceMergeIndex(ForceMergeRequest forceMergeRequest, ActionListener<Void> listener) {
+        assert forceMergeRequest.indices() != null && forceMergeRequest.indices().length == 1 : "DLM force merges one index at a time";
+        final String targetIndex = forceMergeRequest.indices()[0];
+        ActionListener<Void> forceMergeRunningListener = new ActionListener<>() {
+            @Override
+            public void onResponse(Void unused) {
+                logger.trace("DLM issues request to force merge index [{}]", targetIndex);
+                client.admin().indices().forceMerge(forceMergeRequest, new ActionListener<>() {
+                    @Override
+                    public void onResponse(ForceMergeResponse forceMergeResponse) {
+                        logger.info("DLM successfully force merged index [{}]", targetIndex);
+                        setForceMergeMarkerComplete(targetIndex);
+                    }
+
+                    @Override
+                    public void onFailure(Exception e) {
+                        if (e instanceof SnapshotInProgressException) {
+                            logger.info(
+                                "DLM was unable to force merge index [{}] because it's currently being snapshotted. "
+                                    + "Retrying on the next DLM run",
+                                targetIndex
+                            );
+                            // We need to unset the force merge marker in the cluster state so that DLM will try force merging again
+                            unsetForceMergeMarker(targetIndex);
+                        } else {
+                            logger.error(
+                                () -> Strings.format(
+                                    "DLM encountered an error trying to force merge index [%s], and will not retry",
+                                    targetIndex
+                                ),
+                                e
+                            );
+                        }
+                        listener.onFailure(e);
+                    }
+                });
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                listener.onFailure(e);
+            }
+        };
+
+        setForceMergeMarkerBegin(targetIndex, forceMergeRunningListener);
```

Comment:
I'd like to propose we do NOT use a `begin` marker. 

I think this will simplify the design tremendously (i.e. we would not need the `unsetForceMergeMarker` method at all) and I don't think we need it. 

The begin marker is meant to avoid a double execution of forcemerge. But we already have the `resultDeduplicator` to achieve this already (we can just run the same forcemerge request through it on each DLM run and whilst one is in progress the request will not be (re)executed)

As long as a node stays as master, channelling the forcemerge request through the deduplicator will make sure we don't begin the forcemerge request multiple times for the same index. 
On master failover we'll trigger forcemerge again (as we'll have an empty result deduplicator). I think this is fine given we do not forcemerge to 1 segment. 

Even if we keep the `begin forcemerge` cluster marker, on master failover we'll need to re-trigger forcemerge anyway as we wouldn't have any information on the forcemerge operation actually being started (or completed) by the previous master node (we'd just now that there's a marker in the cluster state). 

My proposal here is to simply channel the forcemerge request through the result deduplicator.

I'd say we _could_ keep configuring `end of forcemerge timestamp` value in the cluster state (this will be useful in the future when we add new operations after forcemerge) but at this stage even that could be avoided (if we do decide to keep it I think we should only remove the forcemerge request from the `deduplicator` _after_ we update the cluster state to include the  `end of forcemerge timestamp`)

What do you think? 

@dakrone I'd be interested what you think on this as well

Reply:
Putting in the begin marker came out of my paranoia from seeing some very long-running (multi day) force merges in production. Those were merges to a single segment, and this is not. But I wasn't really sure how the performance would be with the updated force merge that @jpountz is planning on. If it's always going to be minutes or less then what you're saying makes sense to me. If there's a chance it could be hours or days then I'm not as sure.

Reply:
@masseyke I don't think the initial cluster state update to configure the marker would help us more than the `deduplicator` already does - after a master failover just detecting the cluster state maker from a previous master node would not be able to tell us if forcemerge completed - just as with the deduplicator - so we would have to trigger the forcemerge again anyway - just as we'd do had we just used the deduplicator.

Reply:
We definitely have to have _some_ marker in the cluster state right? Unlike a delete (where the index is just gone), there's no way to tell whether an index has been previously force merged. Although maybe there's no harm in force merging an index that has already been force merged on master restart? It would essentially be a no-op, and it might be only a little slower than a cluster state read. I'm leaning toward leaving it though.
I can remove the begin timestamp -- we can always add it back in the future if we need to.

Reply:
++ the end of operation marker (as a timestamp) makes sense, we should not remove the `forcemerge` request from the deduplicator until the cluster state update that sets the end of forcemerge timestamp is completed.